### PR TITLE
[ENH] Categorical power-divergence tests

### PR DESCRIPTION
**Pull request by adam2392**
Towards: #5 

Changes proposed in this pull request:
- categorical (conditional) ind tests

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.
- [ ] I have added a changelog entry succintly describing the change in this PR in the [whats_new](https://github.com/py-why/pywhy-stats/blob/main/docs/whats_new/) relevant version document.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
@bloebp this adds the categorical tests ported from dodiscover. I also ported over the unit-tests and testdata.

The one issue is that the original implementation uses pandas operations like groupby to simplify the contingency table conditioned on the `Z` variable. I leverage pandas as an optional dependency for now to make sure the code performs exactly the same. We can definitely remove that, but rn it's quite complex to do so and involved numpy code, so was wondering if we can table that for the future.

File: pywhy_stats/categorical.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+import numpy as np
+from numpy.typing import ArrayLike
+from scipy import stats
+from sklearn.preprocessing import LabelEncoder
+
+from .pvalue_result import PValueResult
+
+
+def ind(X: ArrayLike, Y: ArrayLike, lambda_: str = "cressie-read") -> PValueResult:
+    """Perform an independence test using power divergence test.
```
Comment (bloebp) [inline pywhy_stats/categorical.py:26]: The overview of the `lambda_` is great, but a bit confusing what this means (without reading the paper). What about renaming it `method` or having a better description how this prefix `lambda_` relates to these other values without needing to read the paper.

File: pywhy_stats/categorical.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
+"""Independence test among categorical variables using power-divergence tests.
```
Comment (bloebp) [inline pywhy_stats/categorical.py:1]: The module name `categorical.py` is quite general. I don't think this will cover all possible categorical tests. What about `power_divergence.py`? Maybe it also makes sense now to have sub-packages like "independence". Then we can have another subpackage like "ksample" etc., depending on the test.

File: pywhy_stats/categorical.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    return _power_divergence(X=X, Y=Y, Z=None, lambda_=lambda_)
+
+
+def condind(
+    X: ArrayLike, Y: ArrayLike, condition_on: ArrayLike, lambda_: str = "cressie-read"
```
Comment (bloebp) [inline pywhy_stats/categorical.py:60]: Same comment as for the `ind` version.

File: pywhy_stats/categorical.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+# This is a modified function taken from pgmpy: License MIT
+def _power_divergence(
+    X: ArrayLike, Y: ArrayLike, Z: Optional[ArrayLike], lambda_: str = "cressie-read"
+) -> PValueResult:
+    """Compute the Cressie-Read power divergence statistic.
+
+    The null hypothesis for the test is X is independent of Y given Z. A lot of the
+    frequency comparison based statistics (eg. chi-square, G-test etc) belong to
+    power divergence family, and are special cases of this test.
```
Comment (bloebp) [inline pywhy_stats/categorical.py:140]: I guess this is the crucial bit that should be part of the public-facing API.

File: pywhy_stats/categorical.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+
+def _compute_conditional_contingency_table(X, Y, Z):
+    unique_Z_vals = np.unique(Z)
+
+    chi = 0
+    dof = 0
+
+    for z_state in unique_Z_vals:
+        pass
```
Comment (bloebp) [inline pywhy_stats/categorical.py:100]: This seems to be a left-over template and is overridden by the next method definition in line 103. Just remove.

File: pywhy_stats/categorical.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        raise ValueError("X and Y should be 1-D arrays.")
+
+    # ensure we use numpy arrays
+    X = np.asarray(X)
+    Y = np.asarray(Y)
+
+    # Check if all elements are integers
+    if not np.issubdtype(X.dtype, np.integer):
+        le = LabelEncoder()
+        X = le.fit_transform(X)
```
Comment (bloebp) [inline pywhy_stats/categorical.py:189]: - We do not need any order here, right? So, "apple" and "pear" would work as inputs?
- What happen if the inputs are continuous floats? Is the idea to catch this in line 202? But then, this would still go through with a small data set (e.g. 30 samples) since then we have 60 unique values. What about throwing an error when one of the variable is a non-discrete or non-categorical value. There are also checks for this.

File: pywhy_stats/categorical.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+    for name, arr in zip(["X", "Y"], [X, Y]):
+        # Check if the number of unique values is reasonably small
+        unique_values = np.unique(arr)
+        num_unique_values = len(unique_values)
+        if num_unique_values > 64:  # Adjust the threshold as needed
+            raise RuntimeError(
+                f"There are {num_unique_values} unique categories for {name}. "
+                f"This is likely an error."
+            )
```
Comment (bloebp) [inline pywhy_stats/categorical.py:206]: Where does this 64 come from? Just a random choice?

File: pywhy_stats/categorical.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        n_samples = Z.shape[0]
+
+        levels_x = len(np.unique(X))
+        levels_y = len(np.unique(Y))
+        dof_check = (
+            (levels_x - 1)
+            * (levels_y - 1)
+            * np.prod([len(np.unique(Z[:, i])) for i in range(s_size)])
+        )
+        n_samples_req = 10 * dof_check
```
Comment (bloebp) [inline pywhy_stats/categorical.py:246]: Where does the `10` come from?

File: pywhy_stats/categorical.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+                c, _, d, _ = stats.chi2_contingency(sub_table_z, lambda_=lambda_)
+                chi += c
+                dof += d
+            except ValueError:
+                # If one of the values is 0 in the 2x2 table.
+                if isinstance(z_state, str):
+                    logging.info(f"Skipping the test X \u27C2 Y | Z={z_state}. Not enough samples")
+                else:
+                    z_str = ", ".join([f"{var}={state}" for var, state in zip(Z_columns, z_state)])
+                    logging.info(f"Skipping the test X \u27C2 Y | {z_str}. Not enough samples")
```
Comment (bloebp) [inline pywhy_stats/categorical.py:277]: This means the test is still running and finishing even if potentially the majority of the `z_state` caused an exception? Would this really lead to the correct result then?

File: tests/test_categorical.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    assert_almost_equal(result.statistic, 289.62, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -139.82, decimal=1)
+    assert result.additional_information["dof"] == 4
+
+    X = df_adult["Immigrant"]
+    Y = df_adult["Sex"]
+    result = categorical.ind(X=X, Y=Y, lambda_=lambda_)
+    assert_almost_equal(result.statistic, 0.2724, decimal=1)
+    assert_almost_equal(np.log(result.pvalue), -0.50, decimal=1)
+    assert result.additional_information["dof"] == 1
```
Comment (bloebp) [inline tests/test_categorical.py:64]: Wondering if we shouldn't only use artificial data here instead? I don't see the real benefit of having real-world data as compared to artificial one in the context of a unit test where we need to know the ground truth.

**Review by bloebp — COMMENTED** [review]
Thanks Adam!
I think having a pandas dependency is even fine as a "hard" requirement. It was maybe important to have it as optional a few years ago, but now, pandas is so common that the use cases where people really want to avoid it is probably non-existence nowadays.

File: tests/test_categorical.py (hunk 10 above)
Comment (adam2392) [reply tests/test_categorical.py:64]: This is to ensure feature parity wrt daggity and pcalg in R. Since those have been relatively vetted, it's a good regression test. I also have artificial data here.

File: pywhy_stats/categorical.py (hunk 7 above)
Comment (adam2392) [reply pywhy_stats/categorical.py:206]: Yes random choice. For a category to have over 64 possible states is a lot. Honestly I think it can be even lower because if we're talking about discrete categorical data, I would be surprised if there is more than 10 states.

If it's an issue w/ users, we can revisit this issue? I can add a `# XXX:` inline comment.

File: pywhy_stats/categorical.py (hunk 2 above)
Comment (adam2392) [reply pywhy_stats/categorical.py:1]: Sure

File: pywhy_stats/categorical.py (hunk 1 above)
Comment (adam2392) [reply pywhy_stats/categorical.py:26]: Changed to `method`.

File: pywhy_stats/categorical.py (hunk 6 above)
Comment (adam2392) [reply pywhy_stats/categorical.py]: Idk if there's an agreed upon method for checking if data is categorical? There's a lot of different types possible: https://numpy.org/doc/stable/reference/arrays.dtypes.html. 

I suppose we can do the naive thing and just check for strings/ints and error out otherwise. But my guess is things like `uint` will fail even though it should work, etc.

Comment (adam2392) [reply pywhy_stats/categorical.py]: I think to ensure a proper sanity check (we can apply this in the KCI test too), we need to basically have some sort of internal contract of what we expect the input arrays to be once they reach the internal computation (e.g. np.int if categorical). I implemented a more complex preprocessing that does a bunch of error checks. 

At the end it will guarantee:

- each element in each array has the same type per column. Otherwise, what does `['test', 1, 1.0, -1, 'other string']` mean? We can document in the docstring that strings are auto-encoded, but otherwise it most likely is some custom preprocessing that the user needs to take.
- if a column of any (X, Y, Z) array is comprised of string subtypes, it will be encoded using a label encoder
- otherwise if it is a type we do not recognize, we throw an error

Comment (bloebp) [reply pywhy_stats/categorical.py]: I believe we don't need this for KCI, as we can adjust the kernel type using a straightforward convention: If it's categorical, encode it as a string. If it's discrete with an order, use integers, and if it's continuous, use floats. In the KCI context, this typically boils down to "if the values in the column are strings, use the delta kernel. Otherwise, default to RBF."

A common point of confusion is that integers or discrete values are not necessarily categorical. Hence, we need an easy way to identify categorical values without requiring the user to know about a specific numpy data type. Given that in many machine learning problems categorical data is already encoded as strings (such as classes, attributes, etc.), simply using the string convention requires the least effort from a user. Most of the time, they don't even need to be aware of this convention.

Regarding `['test', 1, 1.0, -1, 'other string']`, I was thinking about this as well before, but I feel like this is a typical case of "over-engineering", where we don't need to catch all kind of inputs or stuff that can go wrong. Only the less obvious things.

Comment (adam2392) [reply pywhy_stats/categorical.py]: Okay that sounds good

**Comment by adam2392** [comment]
Kay the CI should be fixed w/ this last commit. I'll let you squash-merge if it looks good to you

File: pywhy_stats/categorical.py (hunk 9 above)
Comment (bloebp) [reply pywhy_stats/categorical.py]: Sounds good, can you add this?

File: pywhy_stats/categorical.py (hunk 8 above)
Comment (bloebp) [reply pywhy_stats/categorical.py]: I think used a similar factor in a f-test :) Not sure if exposing it as a parameter makes sense, since a user could just adjust it such that it runs.
What about, we just give out a warning instead of a runtime error. Strictly speaking, it is not an error and also not a hard assumption. We would just expect less accurate results if there are less samples, i.e., it shouldn't cause a runtime error if a user (on their own risk) calls it with too few samples.

**Comment by bloebp** [comment]
Thanks, only two comments left. Then, good to go!

**Comment by adam2392** [comment]
Addressed comments in https://github.com/py-why/pywhy-stats/pull/18/commits/d8ee422bfc25fcace88c2c3dbc5d92a40c254e32

**Review by bloebp — APPROVED** [review]
